project3/coordinator/messenger.h

```diff
@@ -0,0 +1,77 @@
+/**
+ * @File Messenger to be utilized by the coordinator.
+ */
+#ifndef PROJECT3_MESSENGER_H
+#define PROJECT3_MESSENGER_H
+
+#include <pub_sub_messages.pb.h>
+#include <sstream>
+#include <string>
+#include <memory>
+
+#include "connected_participants.h"
+#include "message_log.h"
+#include "pub_sub_messages.pb.h"
+#include "queue/queue.h"
+#include "wire_protocol/wire_protocol.h"
+namespace coordinator::messenger {
+using namespace coordinator::message_log;
+using namespace coordinator::connected_participants;
+/**
+ * @class Handles message broadcasting to participants.
+ *
+ */
+class Messenger {
+ public:
+  /**
+   * @param msg_queue The coordinator's message queue.
+   * @param msg_log The coordinator's message log.
+   * @param connected_participants The coordinator's set of connected
+   * participants.
+   */
+  Messenger(std::shared_ptr<queue::Queue<MessageLog::Message>> msg_queue,
+            std::shared_ptr<MessageLog> msg_log,
+            std::shared_ptr<ConnectedParticipants::Participant> participant,
+            std::shared_ptr<ConnectedParticipants> connected_participants);
+
+  /**
+   * @brief Broadcasts a given message to all active participants.
+   * @note Messages sent by the messenger should be pushed in the message queue before
+   *       this function is called. Guarantees that messages will be broadcast in the order
+   *       they were received.
+   * @return true on success, false on failure
+   *
+   */
+  bool BroadcastMessage();
```

Comment:
It's a little bit of an odd design, IMO, to have this read from a queue internally. Perhaps the queue-reading should happen outside of this class, and this function should simply take the message to broadcast as an argument? That would reduce coupling too, which is always desirable.

Reply:
Agreed. That's what I was thinking too.